prov/verbs/src/verbs_cq.c

```diff
@@ -182,97 +202,124 @@ static uint64_t fi_ibv_comp_flags(struct ibv_wc *wc)
-		entry += 1;
-	}
+static void fi_ibv_cq_read_data_entry(struct ibv_wc *wc, int i, void *buf)
+{
+	struct fi_cq_data_entry *entry = buf;
 
-	return i ? i : (ret ? ret : -FI_EAGAIN);
+	entry[i].op_context = (void *) (uintptr_t) wc->wr_id;
+	entry[i].flags = fi_ibv_comp_flags(wc);
+
+	entry[i].data = (wc->wc_flags & IBV_WC_WITH_IMM) ?
+		ntohl(wc->imm_data) : 0;
+
+	entry->len = (wc->opcode & (IBV_WC_RECV | IBV_WC_RECV_RDMA_WITH_IMM)) ?
+		wc->byte_len : 0;
 }
 
-static ssize_t fi_ibv_cq_read_msg(struct fid_cq *cq, void *buf, size_t count)
+static int fi_ibv_match_ep_id(struct slist_entry *entry, const void *ep_id)
 {
-	struct fi_ibv_cq *_cq;
-	struct fi_cq_msg_entry *entry = buf;
-	ssize_t ret = 0, i;
+	struct fi_ibv_msg_epe *epe = container_of(entry, struct fi_ibv_msg_epe, entry);
 
-	_cq = container_of(cq, struct fi_ibv_cq, cq_fid);
-	if (_cq->wc.status)
-		return -FI_EAVAIL;
+	if (epe->ep->ep_id == (uint64_t) ep_id)
+		return 1;
 
-	for (i = 0; i < count; i++) {
-		ret = ibv_poll_cq(_cq->cq, 1, &_cq->wc);
+	return 0;
+}
+
+ssize_t fi_ibv_poll_cq(struct fi_ibv_cq *cq, struct ibv_wc *wc)
+{
+	struct fi_ibv_msg_epe *epe;
+	struct slist_entry *entry;
+	ssize_t ret;
+
+	while (1) {
+		ret = ibv_poll_cq(cq->cq, 1, wc);
 		if (ret <= 0)
-			break;
+			return ret;
 
-		if (_cq->wc.status) {
-			ret = -FI_EAVAIL;
-			break;
-		}
+		if (wc->opcode == IBV_WC_RECV || wc->opcode == IBV_WC_RECV_RDMA_WITH_IMM)
+			return ret;
 
-		entry->op_context = (void *) (uintptr_t) _cq->wc.wr_id;
-		entry->flags = fi_ibv_comp_flags(&_cq->wc);
-		entry->len = (uint64_t) _cq->wc.byte_len;
-		entry += 1;
-	}
+		if ((wc->wr_id & cq->wr_id_mask) != cq->send_signal_wr_id)
+			return ret;
 
-	return i ? i : (ret ? ret : -FI_EAGAIN);
+		fastlock_acquire(&cq->ep_list_lock);
+		entry = slist_remove_first_match(&cq->ep_list, fi_ibv_match_ep_id, (void *)wc->wr_id);
+		if (!entry) {
+			FI_WARN(&fi_ibv_prov, FI_LOG_CQ, "No matching EP for :"
+					"given signaled send completion\n");
+			fastlock_release(&cq->ep_list_lock);
+			return -FI_EOTHER;
+		}
+		epe = container_of(entry, struct fi_ibv_msg_epe, entry);
+		atomic_sub(&epe->ep->unsignaled_send_cnt,
+				VERBS_SEND_SIGNAL_THRESH(epe->ep));
+		atomic_dec(&epe->ep->comp_pending);
+		fastlock_release(&cq->ep_list_lock);
+		util_buf_release(cq->epe_pool, epe);
+	}
 }
 
-static ssize_t fi_ibv_cq_read_data(struct fid_cq *cq, void *buf, size_t count)
+static ssize_t fi_ibv_cq_read(struct fid_cq *cq_fid, void *buf, size_t count)
 {
-	struct fi_ibv_cq *_cq;
-	struct fi_cq_data_entry *entry = buf;
+	struct fi_ibv_cq *cq;
+	struct fi_ibv_wce *wce;
+	struct slist_entry *entry;
+	struct ibv_wc wc;
 	ssize_t ret = 0, i;
 
-	_cq = container_of(cq, struct fi_ibv_cq, cq_fid);
-	if (_cq->wc.status)
-		return -FI_EAVAIL;
+	cq = container_of(cq_fid, struct fi_ibv_cq, cq_fid);
+
+	fastlock_acquire(&cq->lock);
 
 	for (i = 0; i < count; i++) {
-		ret = ibv_poll_cq(_cq->cq, 1, &_cq->wc);
+		if (!slist_empty(&cq->wcq)) {
```

Comment:
We do not know if wcq would have count entries in it. So this check has to stay inside the loop.